chicken-bug.1
.\" dummy line
.TH CHICKEN-BUG 1 "19 Sep 2001"

.SH NAME

chicken-bug \- generates a bug report from user input

.SH SYNOPSIS

.B chicken-bug
[
.I filename
|
.I OPTION ...
]

.SH OPTIONS

.TP
.B \-help
Show usage information.

.TP
.B \-to-stdout
Write bug report to standard output instead of writing it to a file.

.TP
.B \-
Read user input from standard input, even if files are given on the command line.

.SH DESCRIPTION

.I chicken-bug
generates a bug report file that collects user-input, optional
files and system information obtained from the installed CHICKEN
executables and libraries. Non-option arguments given on the command line
are inserted into the bug report. If no option is given, or if 
.B \-
has been passed on the command line, then a description of the problem
is read from standard input, until EOF (CTRL-D) or a line consisting only
of "." (period) is read.

The bug report will be written to a file in the current directory,
which should be sent to CHICKEN maintainers as it contains various
useful bits of information that make it easier to classify the 
problem described.

Bug reports should be as detailed as possible. It is also very helpful
to include code that reproduces the problem. The more detailed the input
and the more thorough the description is, the easier you make it for
the CHICKEN developers to help you.

.SH BUGS
Submit bug reports by e-mail to
.I chicken-janitors@nongnu.org

.SH AUTHOR
The CHICKEN Team.

.SH SEE ALSO
.BR chicken(1)


NEWS
4.5.2

- added `boot-chicken' makefile target to simplify bootstrapping
  the system from sources and documented this in the README
  (suggested by Jim Ursetto)
- CHICKEN can now be built on haiku (contributed by Chris Roberts)
- On Solaris, the system can be compiled with the SunPro C compiler
  (thanks to Semih Cemiloglu)
- removed the `-disable-warnings' compiler option and `disable-warnings'
  declaration specifier
- `fx/' and `fxmod' generate now faster code in safe mode
- cleaned up manual pages
- slightly optimized variable-value and procedure-access
- in the compiler `-debug-level 2' implies `-scrutinize'
- internal compiler-transformation for `for-each' and `map'
  apply now with any expression as the procedure argument
- the compiler warns about non-intrinsic exported toplevel variables
  which are declared to be safe
- `csc' didn't handle the `-verbose' option (thanks to Mario Goulart)
- the `,d' command in `csi' now detects circular lists (thanks to
  Christian Kellermann)
- strings passed to C runtime functions and which are converted to
  0-terminated C strings are now checked for not containing embedded
  `\0' characters (thanks to Peter Bex)
- errors in user-defined record printers are caught and shown in the 
  output instead of throwing an error to avoid endless recursion when
  an error message is printed
- a feature identifier named `chicken-MAJOR.MINOR' is now defined 
  to simplify conditionalization of code on the CHICKEN version
- `getter-with-setter' copies the lambda-information (if available) from
  the getter to the generated procedure
- `time' uses a somewhat more concise output format and outputs timing
  information to stderr instead of stdout
- added a new chapter on cross-development to the manual
- added the `safe-globals' declaration specifier
- split up manual chapter `Modules and macros' into two chapters
  (named `Modules' and `Macros', respectively - suggested by
  Mario Goulart)
- the last 5 non-precompiled regular expressions are now internally 
  cached in compiled form, which speeds up repeated matching of
  the same uncompiled regular expression dramatically
- added the new procedure `yes-or-no?' to the `utils' library unit
- added a `bench' makefile target that runs some non-trivial
  benchmark programs

4.5.1

- added `install-target' and `install-dev' makefile target for
  installing only target system libraries in cross-mode and
  development files (static libraries and headers)
- added `[-]no-procedure-checks-for-toplevel-bindings' compiler
  option and declaration
- usage of unimported syntax in modules gives more usable
  error messages; in particular, used but unimported FFI forms are 
  now detected
- invalid syntactic forms (mostly `()') encountered by the compiler
  or interpreter show the contextual form containing the expression,
  or, if indicated by the context warns about missing imports
- simplified manual pages of all core tools - they now refer to
  the output shown by invoking `<tool> -help'
- added new option `-feature FEATURE' to `chicken-install' tool
  to pass feature-identifiers to invocations of `csc'
- removed deprecated `-host-extension' option from `chicken-install'
- `chicken-status' in a system built for cross-compilation now 
  lists extensions installed in the target prefix, unless the
  new `-host' option is given on the command line
- `chicken-uninstall' in a system built for cross-compilation now 
  removes extensions installed in the target prefix, unless the
  new `-host' option is given on the command line
- added missing entry for `finite?' to the `chicken' module
  exports
- added new procedure `port-closed?' to the `library' unit
  (contributed by Peter Bex)
- added new procedure `symbol-append' to the library unit
- the compiler-option `-optimize-level 0' is equivalent to
  `-no-compiler-syntax -no-usual-integrations`
- internal rewritings of `map' and `for-each' ensure correct
  evaluation order of the arguments and does a better job
  at detecting non-list arguments (thanks to Jim Ursetto)
- `void' now takes arbitrary arguments and ignores them
- deprecated `noop' (from the `data-structures' unit) which
  is now replaced by `void'
- the `time' macro now performs a major garbage collection
  before executing the contained expressions and writes
  the timings in a more compact format to the port given
  by `(current-error-port)' instead of the standard output
  port
- definitions of the form `(define define ...)' and
  `(define-syntax define-syntax ...)' now trigger an error,
  as required by R5RS (thanks to Jeronimo Pellegrini and Alex
  Shinn)
- deprecated `random-seed' from the `extras' unit, since it
  is identical to `randomize'
- added new procedure `create-temporary-directory' to the
  `files' unit
- deprecated the optional path separator argument to
  `make-pathname'
- slightly improved the performance of keyword argument
  processing
- the second argument to `find-files' from the `posix' unit
  is now optional and defaults to `(constantly #t)'
- removed the deprecated `canonical-path' and `current-environment'
  procedures from the `posix' unit
- warnings that mostly refer to programming style issues are
  now coined `notes' and are only shown in the interpreter
  or when debug-mode is enabled or when scrutiny is enabled
  when compiling

4.5.0

- internal fixes of handling of alternative installation-prefix
  in setup-api
- certain compiler-warnings that are in really just notes
  and don't indicate a possible error (like reimport of
  identifiers) are only shown with -S or in verbose mode
- fixed handling of VARDIR in `chicken-install' (thanks to
  Davide Puricelli)
- `chicken-install -test' doesn't runs tests for dependencies
- when a non-else clause follows an else-clause in `cond', 
  `case' or `select' a warning (note) is shown in verbose
  mode
- removed the deprecated `define-extension' and
  `define-compiled-syntax'
- `chicken-uninstall' now always asks before removing 

chicken-bug.scm
;;;; chicken-bug.scm - Bug report-generator
;
; Copyright (c) 2008-2010, The Chicken Team
; All rights reserved.
;
; Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following
; conditions are met:
;
;   Redistributions of source code must retain the above copyright notice, this list of conditions and the following
;     disclaimer. 
;   Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following
;     disclaimer in the documentation and/or other materials provided with the distribution. 
;   Neither the name of the author nor the names of its contributors may be used to endorse or promote
;     products derived from this software without specific prior written permission. 
;
; THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
; AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR
; CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
; CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
; SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
; THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
; OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
; POSSIBILITY OF SUCH DAMAGE.


(require-extension srfi-13 posix tcp data-structures utils extras)


(define-constant +bug-report-file+ "chicken-bug-report.~a-~a-~a")

(define-constant +fallbackdestinations+ 
  "chicken-janitors@nongnu.org\nchicken-hackers@nongnu.org\nchicken-users@nongnu.org")

(define-constant +destination+ "chicken-janitors@nongnu.org")
(define-constant +mxservers+ '() ; XXX temporarily disabled 
  #;(list "mx10.gnu.org" "mx20.gnu.org"))
(define-constant +send-tries+ 3)

(define-foreign-variable +cc+ c-string "C_TARGET_CC")
(define-foreign-variable +cxx+ c-string "C_TARGET_CXX")
(define-foreign-variable +c-include-path+ c-string "C_INSTALL_INCLUDE_HOME")

(define (user-id)
  (cond-expand
   ((or mingw32 msvc) "<not available>")
   (else (user-information (current-user-id)))))

(define (collect-info)
  (print "\n--------------------------------------------------\n")
  (print "This is a bug report generated by chicken-bug(1).\n")
  (print "Date:\t" (seconds->string (current-seconds)) "\n\n")
  (printf "User information:\t~s~%~%" (user-id))
  (print "Host information:\n")
  (print "\tmachine type:\t" (machine-type))
  (print "\tsoftware type:\t" (software-type))
  (print "\tsoftware version:\t" (software-version))
  (print "\tbuild platform:\t" (build-platform) "\n")
  (print "CHICKEN version is:\n" (chicken-version #t) "\n")
  (print "Home directory:\t" (chicken-home) "\n")
  (printf "Include path:\t~s~%~%" ##sys#include-pathnames)
  (print "Features:")
  (for-each
   (lambda (lst) 
     (display "\n  ")
     (for-each 
      (lambda (f)
	(printf "~a~a" f (make-string (fxmax 1 (fx- 16 (string-length f))) #\space)) )
      lst) )
   (chop (sort (map keyword->string ##sys#features) string<?) 5))
  (print "\n\nchicken-config.h:\n")
  (with-input-from-file (make-pathname +c-include-path+ "chicken-config.h")
    (lambda ()
      (display (read-all)) ) )
  (newline)
  (when (and (string=? +cc+ "gcc") (feature? 'unix))
    (print "CC seems to be gcc, trying to obtain version...\n")
    (with-input-from-pipe "gcc -v 2>&1"
      (lambda ()
	(display (read-all)))))
  (newline) )

(define (usage code)
  (print #<<EOF
usage: chicken-bug [FILENAME ...]

  -help  -h            show this message
  -to-stdout           write bug report to standard output
  -                    read description from standard input

Generates a bug report file from user input or alternatively
from the contents of files given on the command line.

EOF
) 
  (exit code) )

(define (user-input)
  (when (##sys#tty-port? (current-input-port))
    (print #<<EOF
This is the CHICKEN bug report generator. Please enter a detailed
description of the problem you have encountered and enter CTRL-D (EOF)
or a line consisting only of "." to finish. Press CTRL-C to abort the program. You can
also pass the description from a file (just abort now and re-invoke
"chicken-bug" with one or more input files given on the command-line)

EOF
) )
  (let loop ((data '()))
    (let ((ln (read-line)))
      (cond ((or (eof-object? ln) (string=? "." ln))
	     (string-concatenate-reverse data) )
	    (else (loop (cons ln data)))))))

(define (justify n)
  (let ((s (number->string n)))
    (if (> (string-length s) 1)
	s
	(string-append "0" s))))

(define (main args)
  (let ((msg "")
	(files #f)
	(stdout #f))
    (for-each
     (lambda (arg)
       (cond ((string=? "-" arg) 
	      (set! files #t)
	      (set! msg (string-append msg "\n\nUser input:\n\n" (user-input))) )
	     ((member arg '("--help" "-h" "-help"))
	      (usage 0) )
	     ((string=? "-to-stdout" arg)
	      (set! stdout #t) )
	     (else
	      (set! files #t)
	      (set! msg 
		(string-append
		 msg
		 "\n\nFile added: " arg "\n\n"
		 (read-all arg) ) ) ) ) )
     args)
    (unless files
      (set! msg (string-append msg "\n\n" (user-input))))
    (newline)
    (let* ((lt (seconds->local-time (current-seconds)))
	   (day (vector-ref lt 3))
	   (mon (vector-ref lt 4))
	   (yr (vector-ref lt 5)) )
      (if stdout
	  (begin
	    (print msg)
	    (collect-info))
	  (try-mail
	   +mxservers+
	   (sprintf +bug-report-file+ (+ 1900 yr) (justify mon) (justify day))
	   (mail-headers)
	   (with-output-to-string
	     (lambda ()
	       (print msg)
	       (collect-info))))))))
      ;(let* ((file (sprintf +bug-report-file+ (+ 1900 yr) (justify mon) (justify day)))
	;     (port (if stdout (current-output-port) (open-output-file file))))
	;(with-output-to-port port
	;  (lambda ()
	;    (print msg)
	;    (collect-info) ) )
	;(unless stdout
	;  (close-output-port port)
	;  (print "\nA bug report has been written to `" file "'. Please send it to")
	;  (print "one of the following addresses:\n\n" +destinations+) ) ) ) ) )

(define (try-mail servs fname hdrs msg)
    (if (null? servs)
        (begin
            (with-output-to-file fname
                (lambda () (print msg)))
            ;XXX temporarily disabled:
	    ;(print "\nCould not send mail automatically!\n")
	    (print "\nA bug report has been written to `" fname "'.  Please send it to")
            (print "one of the following addresses:\n\n" +fallbackdestinations+))
        (or (send-mail (car servs) msg hdrs fname)
            (try-mail (cdr servs) fname hdrs msg))))

(define (mail-date-str tm)
    (string-append
        (case (vector-ref tm 6)
            ((0) "Sun, ")
            ((1) "Mon, ")
            ((2) "Tue, ")
            ((3) "Wed, ")
            ((4) "Thu, ")
            ((5) "Fri, ")
            ((6) "Sat, "))
        (string-pad (number->string (vector-ref tm 3)) 2 #\0)
        (case (vector-ref tm 4)
            ((0)  " Jan ")
            ((1)  " Feb ")
            ((2)  " Mar ")
            ((3)  " Apr ")
            ((4)  " May ")
            ((5)  " Jun ")
            ((6)  " Jul ")
            ((7)  " Aug ")
            ((8)  " Sep ")
            ((9)  " Oct ")
            ((10) " Nov ")
            ((11) " Dec "))
        (number->string (+ 1900 (vector-ref tm 5)))
        " "
        (string-pad (number->string (vector-ref tm 2)) 2 #\0)
        ":"
        (string-pad (number->string (vector-ref tm 1)) 2 #\0)
        ":"
        (string-pad (number->string (vector-ref tm 0)) 2 #\0)
        " +0000"))

(define (mail-headers)
    (string-append
        "Date: " (mail-date-str (seconds->utc-time (current-seconds))) "\r\n"
        "From: \"chicken-bug user\" <chicken-bug-command@callcc.org>\r\n"
        "To: \"Chicken Janitors\" <chicken-janitors@nongnu.org>\r\n"
        "Subject: Automated chicken-bug output -- "))

(define (mail-read i o)
    (let ((v   (condition-case (read-line i)
                   (var () (close-input-port i) (close-output-port o) #f))))
        (if v
            (if (char-numeric? (string-ref v 0))
                (string->number (substring v 0 3))
                (mail-read i o))
            #f)))

(define (mail-write i o m)
    (let ((v   (condition-case (display m o)
                   (var () (close-input-port i) (close-output-port o) #f))))
        (if v
            (mail-read i o)
            #f)))

(define (mail-check i o v e k)
    (if (and v (= v e))
        #t
        (begin
            (close-input-port i)
            (close-output-port o)
            (k #f))))

(define (send-mail serv msg hdrs fname)
  (call/cc 
   (lambda (return)
     (do ((try 1 (add1 try)))
	 ((> try +send-tries+))
       (print* "connecting to " serv ", try #" try " ...")
       (receive (i o)
	   (tcp-connect serv 25)
	 (call-with-current-continuation
	  (lambda (k)
	    (mail-check i o (mail-read i o) 220 k)
	    (mail-check i o (mail-write i o "HELO callcc.org\r\n") 250 k)
	    (mail-check i o (mail-write i o "MAIL FROM:<chicken-bug-command@callcc.org>\r\n") 250 k)
	    (mail-check i o (mail-write i o "RCPT TO:<chicken-janitors@nongnu.org>\r\n") 250 k)
	    (mail-check i o (mail-write i o "DATA\r\n") 354 k)
	    (mail-check i o (mail-write i o (string-append hdrs fname "\r\n\r\n" msg "\r\n.\r\n")) 250 k)
	    (display "QUIT" o)
	    (close-input-port i)
	    (close-output-port o)
	    (print "ok.\n\nBug report successfully mailed to the Chicken maintainers.\nThank you very much!\n\n")
	    (return #t))))
       (print " failed.")))))

(main (command-line-arguments))
